Improve time management

XInTheDark opened the pull request:
Introduce new time management formulas. The time calculated is a hard upper limit, and is passed to search in the same way as a movetime limit. Most of the parameters were handpicked, so tuning them should bring some more gains.

Passed STC: https://montychess.org/tests/view/668c8b23ca41bb3d0a8fac19
LLR: 2.92 (-2.94,2.94) <0.00,4.00>
Total: 1920 W: 605 L: 438 D: 877
Ptnml(0-2): 24, 174, 422, 291, 49 

Passed LTC: https://montychess.org/tests/view/668cac44ca41bb3d0a8fac21
LLR: 2.95 (-2.94,2.94) <1.00,5.00>
Total: 3252 W: 818 L: 662 D: 1772
Ptnml(0-2): 22, 306, 832, 426, 40 

Bench: 2149456

XInTheDark committed:
Take 5 rebased
Bench: 2149456

XInTheDark committed:
Take 5r, remove mtg adjust thing
Bench: 2149456

XInTheDark committed:
Take 5r, remove mtg adjust thing, fixed
Bench: 2149456

jw1912 commented on src/chess/board.rs:
```diff
@@ -17,10 +17,11 @@ pub struct Board {
     enp_sq: u8,
     rights: u8,
     halfm: u8,
+    ply: u16,
```
Can you move this out of the board struct, as you mentioned it can be done entirely in the UCI code?

XInTheDark replied:
Done in 0d45487

XInTheDark committed:
Run rustfmt
Bench: 2149456

jw1912 commented on src/mcts/helpers.rs:
```diff
@@ -60,11 +60,40 @@ impl SearchHelpers {
+                i as u64
+            } else {
+                0
+            };
```
```suggestion
        let inc = increment.unwrap_or(0);
```

jw1912 commented on src/mcts/helpers.rs:
```diff
@@ -60,11 +60,40 @@ impl SearchHelpers {
+            max_time = (opt_scale * bonus * time_left) as u128;
+        } else {
+            max_time = (time / mtg) as u128;
+        }
```
```suggestion
        let mut max_time = if let Some(mtg) = movestogo {
            time / mtg.clamp(1, 30)
        } else {
            let mtg = 30;
            
            let time_left = (time + inc * (mtg - 1) - 10 * (2 + mtg)).max(1) as f64;
            let log_time = (time_left / 1000.0).log10();
            
            let opt_constant = (0.0048 + 0.00032 * log_time).min(0.0060);
            let opt_scale = (0.0125 + (ply as f64 + 2.5).sqrt() * opt_constant).min(0.25 * time as f64 / time_left);
            
            // More time at the start of the game
            let bonus = if ply <= 10 { 1.0 + (11.0 - ply as f64).log10() * 0.5 } else { 1.0 };
            
            opt_scale * bonus * time_left
        } as u128;
```

XInTheDark committed:
Refactor
Bench: 2149456

XInTheDark committed:
Remove ply, handle in UCI instead

Note that the ply from FENs isn't parsed, since almost all game managers will pass by moves, not FEN.

Bench: 2149456

XInTheDark committed:
initialise variable
Bench: 2149456

XInTheDark committed:
refactor 2
Bench: 2149456

jw1912 reviewed (approved):
Very nice, hope to see good results from the tune.
